momento-provider/src/main/java/com/google/code/ssm/providers/momento/MomentoClientWrapper.java

```diff
@@ -217,11 +197,47 @@ public Object getNativeClient() {
         return transcoderAdapter;
     }
 
-    private static int getFlag(byte[] data, int i) {
-        return (data[i] & 0xff) << 24
-                | (data[i + 1] & 0xff) << 16
-                | (data[i + 2] & 0xff) << 8
-                | (data[i + 3] & 0xff);
+    // Helper function that concatenates the CachedObject to raw bytes before calling `set`.
+    // This helps maintain our serialization flags so that we can deserialize the object upon a `get`
+    private boolean writeOutToMomento(final String key, final int exp, final int flagsUsed, final byte[] data) {
+        ByteBuffer buffer = ByteBuffer.allocate(SIZE_OF_INT + data.length);
+        buffer.putInt(flagsUsed);
+        buffer.put(data);
+        buffer.rewind();
+        CacheSetResponse response = momentoClient.set(defaultCacheName, key, buffer, exp);
+        return response != null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object
+    private <T> T readFromMomento(final String key, final CacheTranscoder transcoder) {
+        Transcoder<T> cacheTranscoder = getTranscoder(transcoder);
+        Optional<CachedData> maybeCachedData = performGet(key);
+        return maybeCachedData.map(cacheTranscoder::decode)
+                .orElse(null);
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object, uses our default Transcoder
+    private Object readFromMomento(final String key) {
+        CacheTranscoder cacheTranscoder = getTranscoder();
+        Optional<CachedData> maybeCachedData = performGet(key);
+        return maybeCachedData.map(cachedData -> cacheTranscoder.decode(new CachedObjectWrapper(cachedData)))
+                .orElse(null);
+    }
+
+    private Optional<CachedData> performGet(final String key) {
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            ByteBuffer byteBuffer = ByteBuffer.wrap(cacheGetResponse.get());
+            int flags = byteBuffer.getInt();
+            byte[] originalData = byteBuffer.array();
```

Comment:
I think this will break your tests if you run them again (this a good reason to check in tests and run them on PR)

If you want to copy the remaining bytes from an NIO ByteBuffer you want to do
```
// the abstraction here does not support nio types and java does not have pointers so we have
// no choice but to copy the object.
byte[] originalData = new byte[byteBuffer.remaining()];
byteBuffer.get(originalData);
```

Reply:
It certainly didn't break my tests when I ran them, but I'm happy to include it this way to be sure!

Reply:
per https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/ByteBuffer.html#array() this should return `cacheGetResponse.get()`'s value directly (i.e., inclusive of the int prefix), no? Isn't byteBuffer.array() == cacheGetResponse.get()`?